### PR TITLE
Rails 6.1

### PR DESCRIPTION
**Pull request by slawosz**
App updated to rails 6.1

### CODE REVIEW

File: cosmetics-web/app/controllers/submit_application_controller.rb [hunk 1]
```diff
@@ -26,4 +26,9 @@ def try_to_finish_account_setup
   def authorize_user!
     redirect_to invalid_account_path if current_user && !current_user.is_a?(SubmitUser)
   end
+
+  def root_path
+    submit_root_path
+  end
+  helper_method :root_path
```
Comment (scruti) [inline cosmetics-web/app/controllers/submit_application_controller.rb:33]: :question: 
Do we need this override when `application_controller` has this?:
```ruby
  def root_path
    submit_domain? ? submit_root_path : search_root_path
  end
  ```

File: cosmetics-web/app/controllers/responsible_persons/formulation_upload_controller.rb [hunk 2]
```diff
@@ -11,7 +11,7 @@ def create
       if @component.save
         redirect_to responsible_person_notification_additional_information_index_path(@component.notification.responsible_person, @component.notification)
       else
-        @component.formulation_file.purge
+        @component.formulation_file.delete
```
Comment (scruti) [inline cosmetics-web/app/controllers/responsible_persons/formulation_upload_controller.rb:14]: Question :question: 
I am confused about this change. ActiveStorage guides does not have reference to `#delete` anywhere ( https://edgeguides.rubyonrails.org/active_storage_overview.html#removing-files ). But seems you tested this and it actually purges the blob according to the specs... is `delete` an alias? Does something different than `purge`? :man_shrugging:

File: cosmetics-web/app/helpers/component_build_helper.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 module ComponentBuildHelper
   def cmr_errors(component)
-    component.cmrs.each_with_index.flat_map { |cmr, index| cmr.errors.map { |attr, msg| { text: msg, href: "#component_cmrs_attributes_#{index}_#{attr}" } } } +
-      component.errors.reject { |attr, _| attr.to_s.include? "cmrs." }.map { |_, msg| { text: msg, href: "#component_cmrs_attributes_0_name" } }
+    component.cmrs.each_with_index.flat_map { |cmr, index| cmr.errors.map { |error| { text: error.message, href: "#component_cmrs_attributes_#{index}_#{error.attribute}" } } } +
+      component.errors.reject { |_error| attribute.to_s.include? "cmrs." }.map { |error| { text: error.message, href: "#component_cmrs_attributes_0_name" } }
```
Comment (scruti) [inline cosmetics-web/app/helpers/component_build_helper.rb:4]: :raising_hand_man:  
Not 100% sure on this but better safe than sorry. From what I see in the changes now the interface is `error#attribute`  (as used in line 3). I belive the line 4 could be wrong then:
```suggestion
      component.errors.reject { |error| error.attribute.to_s.include? "cmrs." }.map { |error| { text: error.message, href: "#component_cmrs_attributes_0_name" } }
```

File: cosmetics-web/app/helpers/nanomaterial_helper.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 module NanomaterialHelper
   def get_nanomaterial_purpose_options
-    NanoElement.purposes.map { |purpose| [purpose, get_label_for_nanomaterial_purpose(purpose)] }.to_h
+    NanoElement.purposes.index_with { |purpose| get_label_for_nanomaterial_purpose(purpose) }
```
Comment (scruti) [inline cosmetics-web/app/helpers/nanomaterial_helper.rb:3]: Nice <3

File: cosmetics-web/app/helpers/component_build_helper.rb (hunk 3 above)
Comment (slawosz) [reply cosmetics-web/app/helpers/component_build_helper.rb:4]: Good catch!

File: cosmetics-web/spec/requests/secondary_authentication/sms_spec.rb [hunk 5]
```diff
@@ -19,17 +19,21 @@
...
-        expect(notify_stub).to have_received(:send_sms).with(
-          hash_including(phone_number: user.mobile_number, personalisation: { code: user.reload.direct_otp }),
-        )
+
+        perform_enqueued_jobs
+
+        expect(notify_stub).to have_received(:send_sms)
+        #   .with(
+        #   hash_including(phone_number: user.mobile_number, personalisation: { code: user.reload.direct_otp }),
+        # )
```
Comment (scruti) [inline cosmetics-web/spec/requests/secondary_authentication/sms_spec.rb:36]: Just highlighting this for attention. Was this temporal?

**Review by scruti — APPROVED** [review]
LGTM!

**Comment by scruti** [comment]
@slawosz there is an error (triggered by Nasir testing on review app) due to a missing gem: 
https://sentry.io/organizations/beis/issues/2380248575/?project=1398436

**Review by nasirkhanbeis — APPROVED** [review]
Tested this on review app 2088, looking good as expected. 
Verified all journeys, manual and zip upload, validated file uploads, validation errors, create user, invite user etc. all good!

**Review by scruti — CHANGES_REQUESTED** [review]
Please change: Noticed an exception due to a missing gem for image processing.
https://sentry.io/organizations/beis/issues/2380248575/?project=1398436

**Review by scruti — APPROVED** [review]

**Pull request merged**